### PR TITLE
feature/java generics improvements

### PR DESCRIPTION
**Pull request by baywet**
this PR improves the java generation to take advantage of better genericity of the infrastructure and refactors duplicated code.
example changes: https://github.com/microsoftgraph/msgraph-sdk-java/pull/119

### CODE REVIEW

File: Templates/Java/requests_extensions/BaseEntityCollectionReferenceRequest.java.tt [hunk 1]
```diff
@@ -23,8 +29,8 @@ import <#=mainNamespace#>.<#=c.GetPackagePrefix()#>.<#=c.TypeName()#>;
      * @param client         the service client
      * @param requestOptions the options for this request
      */
-    public <#=c.TypeCollectionReferenceRequest()#>(final String requestUrl, <#=TypeHelperJava.IBaseClientType()#> client, final java.util.List<? extends <#=importNamespace#>.options.Option> requestOptions) {
-        super(requestUrl, client, requestOptions, <#=c.TypeCollectionResponse()#>.class, <#=c.TypeCollectionPage()#>.class);
+    public <#=c.TypeCollectionReferenceRequest()#>(final String requestUrl, <#=TypeHelperJava.IBaseClientType()#> client, final java.util.List<? extends <#=mainNamespace#>.options.Option> requestOptions) {
```
Comment (baywet) [inline Templates/Java/requests_extensions/BaseEntityCollectionReferenceRequest.java.tt:32]:
```suggestion
    public <#=c.TypeCollectionReferenceRequest()#>(final String requestUrl, <#=TypeHelperJava.IBaseClientType()#> client, final java.util.List<? extends <#=importNamespace#>.options.Option> requestOptions) {
```

**Comment by baywet** [comment]
The changes of this generation have been run against raptor here.
https://microsoftgraph.visualstudio.com/Graph%20Developer%20Experiences/_build/results?buildId=31836&view=results

It's complaining about the todo stuff as I haven't updated the feature/v3 branch from latest metadata in a while.

**Comment by zengin** [comment]
I have done a second pass on this PR. Nothing stands out to me other than what we have discussed in the Teams meeting.

Sharing my comments from the meeting here too:

Just to make sure that the transfer of methods from child to base types happened as expected, please consider providing a type summary diff.

We do this as part of the .NET generation pipeline
- [Script](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/dev/scripts/generateTypeSummary.ps1)
- [Pipeline](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/dev/.azure-pipelines/generate-v1.0-sdk.yml#L150-L157)

The script loads the built assembly and walks through all the types and lists methods/properties declared directly in that type. It also sorts the results for easy-to-handle diffs.

For this particular case, I think we don't want the filtering for the directly declared methods/properties, to make sure that by inheritance we get the same set after the change.

Other than that, the change looks good. I really appreciate the smaller binary size and added javadocs.

**Comment by MIchaelMainer** [comment]
**Part 1**

I've tested the following calls with these changes to increase confidence in these changes:
- [x] get me
- [x] get users collection and access a user in the current page
- [x] find meeting times

- [x] to test the use of [BaseActionCollectionRequestBuilder.java](https://github.com/microsoftgraph/msgraph-sdk-java/pull/119/files#diff-ecd217ed450b52e6702e4ab34de98e76d609ea2cf68c603f83116e7251f05014)
The good news is that BaseActionCollectionRequestBuilder works well for supporting the generated changes. I did find that EnumSet is not serialized appropriately for this scenario. The following code: 

```java
public void testGetMailtips() {
		TestBase testBase = new TestBase();
		ArrayList<String> users = new ArrayList<String>();
		users.add("michael@chambele.onmicrosoft.com");
		users.add("AdeleV@chambele.onmicrosoft.com");
		EnumSet<MailTipsType> mailtips = EnumSet.of(MailTipsType.MAILBOX_FULL_STATUS, MailTipsType.MAX_MESSAGE_SIZE);
		UserGetMailTipsCollectionPage page = testBase.graphClient.me().getMailTips(users, mailtips).buildRequest().post();
		assertNotNull(page);
	}
```
results in the following body:
```json
{
    "emailAddresses": ["michael@chambele.onmicrosoft.com", "AdeleV@chambele.onmicrosoft.com"],
    "mailTipsOptions": "MAILBOX_FULL_STATUS,MAX_MESSAGE_SIZE"
}
```
expected:
```json
{
    "emailAddresses": ["michael@chambele.onmicrosoft.com", "AdeleV@chambele.onmicrosoft.com"],
    "mailTipsOptions": "mailboxFullStatus,maxMessageSize"
}
```

**Comment by MIchaelMainer** [comment]
**Part 2**

Function bound to a collection that returns a collection.
<Function Name="delta" IsBound="true">
   <Parameter Name="bindingparameter" Type="Collection(graph.mailFolder)" Nullable="false" />
   <ReturnType Type="Collection(graph.mailFolder)" Nullable="false" />
</Function>

We don't generate MailFolderDeltaCollectionRequest to pass down the generic response type to BaseCollectionRequest. This means that `testBase.graphClient.me().mailFolders().delta().buildRequest().get()` returns T3 instead of MailFolderDeltaCollectionPage.

**Comment by MIchaelMainer** [comment]
**Part 3 - todo**

- [x] validate that CollectionPage can be read.
- [ ] exercise entity collection reference request (BaseEntityCollectionReferenceRequest)
- [ ] we properly emit a query parameter in request (validate add*Option)
- [ ] exercise entity collection request (BaseEntityCollectionRequest and BaseEntityCollectionRequestBuilder)
- [ ] exercise reference collection page (BaseEntityCollectionWithReferencesPage).
- [ ] exercise entity collection with references request (BaseEntityCollectionWithReferencesRequest and BaseEntityCollectionWithReferencesRequestBuilder)
- [ ] exercise entity reference request (BaseEntityReferencesRequest and BaseEntityReferencesRequestBuilder)
- [ ] exercise entity with reference request (BaseEntityWithReferenceRequest and BaseEntityWithReferenceRequest)
- [x] exercise method body request (tested with getMailTips)
- [ ] validate that deltaLinks are available on collection page results
- [ ] validate that functions work as expected (look at another example besides Collection(mailfolder).Delta() in **Part 2** since they were moved out of BaseMethodCollectionRequest.

**Review by zengin — APPROVED** [review]
Thank you very much for adding the type summary. It helped a lot. LGTM in terms of typing changes. I will defer to Michael's tests for any potential functionality changes.

**Comment by MIchaelMainer** [comment]
@baywet Does e54ebce addresses the NullPointerException?

**Comment by baywet** [comment]
@MIchaelMainer no it doesn't, I still have to go through the remaining items you provided with the checklist. Thanks for the amazing testing work BTW!

**Comment by baywet** [comment]
For clarity I'm moving the to-dos from the review here:

- [x] Investigate The *CollectionWithReferencesPage looks fine. I encounter a NullPointerException when I make a request to get references: fixed
- [ ] Investigate DirectoryObjectCollectionWithReferencesPage may be an issue when following $ref next pages as it may not instantiate the correct request object.
- [ ] Investigate The *WithReferenceRequest does not contain a definition for GET. It only has a definition for DELETE and PUT. This may be the current state for the library.

**Comment by baywet** [comment]
@MIchaelMainer I've just finished addressing the final 3 points. Let me know whether I missed anything, but this should be good for final review.

**Review by MIchaelMainer — APPROVED** [review]

**Pull request merged**